pkg/apis/validation/validation_test.go

```diff
@@ -237,3 +311,137 @@ func Test_keyEncodingValue(t *testing.T) {
 		})
 	}
 }
+
+func Test_PKCS12Values(t *testing.T) {
+	basePath := field.NewPath("root")
+
+	tests := map[string]struct {
+		attr   map[string]string
+		expErr field.ErrorList
+	}{
+		"if no attributes, expect no error": {
+			attr:   map[string]string{},
+			expErr: nil,
+		},
+		"if key and password is defined, but enabled is not defined, expect error": {
+			attr: map[string]string{
+				"csi.cert-manager.io/keystore-pkcs12-file":     "my-file",
+				"csi.cert-manager.io/keystore-pkcs12-password": "password",
+			},
+			expErr: field.ErrorList{
+				field.Invalid(basePath.Child("csi.cert-manager.io/keystore-pkcs12-file"), "my-file",
+					"cannot use attribute without \"csi.cert-manager.io/keystore-pkcs12-enable\" set to \"true\" or \"false\""),
+				field.Invalid(basePath.Child("csi.cert-manager.io/keystore-pkcs12-password"), "password",
+					"cannot use attribute without \"csi.cert-manager.io/keystore-pkcs12-enable\" set to \"true\" or \"false\""),
+			},
+		},
+
+		"if key and password is defined, and enabled is defined as false, expect no error": {
+			attr: map[string]string{
+				"csi.cert-manager.io/keystore-pkcs12-enable":   "false",
+				"csi.cert-manager.io/keystore-pkcs12-file":     "my-file",
+				"csi.cert-manager.io/keystore-pkcs12-password": "password",
+			},
+			expErr: nil,
+		},
+		"if key and password is defined, but enabled is defined as foo, expect error": {
+			attr: map[string]string{
+				"csi.cert-manager.io/keystore-pkcs12-enable":   "foo",
+				"csi.cert-manager.io/keystore-pkcs12-file":     "my-file",
+				"csi.cert-manager.io/keystore-pkcs12-password": "password",
+			},
+			expErr: field.ErrorList{
+				field.NotSupported(basePath.Child("csi.cert-manager.io/keystore-pkcs12-enable"), "foo", []string{"true", "false"}),
+			},
+		},
+		"if key and password is not defined, and enabled is defined as true, expect error": {
+			attr: map[string]string{
+				"csi.cert-manager.io/keystore-pkcs12-enable": "true",
+			},
+			expErr: field.ErrorList{
+				field.Required(basePath.Child("csi.cert-manager.io/keystore-pkcs12-file"), "required attribute when PKCS12 KeyStore is enabled"),
+				field.Required(basePath.Child("csi.cert-manager.io/keystore-pkcs12-password"), "required attribute when PKCS12 KeyStore is enabled"),
+			},
+		},
+		"if key and password is defined as empty string, and enabled is defined as true, expect error": {
+			attr: map[string]string{
+				"csi.cert-manager.io/keystore-pkcs12-enable":   "true",
+				"csi.cert-manager.io/keystore-pkcs12-file":     "",
+				"csi.cert-manager.io/keystore-pkcs12-password": "",
+			},
+			expErr: field.ErrorList{
+				field.Required(basePath.Child("csi.cert-manager.io/keystore-pkcs12-file"), "required attribute when PKCS12 KeyStore is enabled"),
```

Comment:
This isn't required though as we default it, right? Or is this validation applied after defaulting and the check is there for the sake of completeness?

Also, `KeyStore` vs `keystore` - I think this is the first time I've seen the capitalised version of the word in this PR

Reply:
I _think_ you answered your above questions with your comment here?

Yep, defaulting and validating are distinct functions where validation occurs after defaulting. Being distinct operations, yes there may be impossible states either function accounts for, but implements them for completeness. 

I have also made an effort to default all values (where appropriate, i.e. PKCS12 being enabled), regardless of whether that would be an empty string or not.

--- 

The camel case here is to keep consistency with the [oracle documentation on pkcs12](https://docs.oracle.com/cd/E19509-01/820-3503/ggfhb/index.html). We _should_ have camel case whenever there is human readable text like comments or error messages, and then all lower case for volume attributes which is consistent with DNS names.

Let me know if you find inconsistencies with these two scenarios.